File: src/linear_controller_py/src/linear_controller.py
#!/usr/bin/env python

import rospy

from std_msgs.msg import Float64
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry

##
# Parameters
##
P_GAIN = 40


####################
# Controller class #
####################
class LinearController:
    # Constructor
    def __init__(self, p_gain):

        # Assign controller parameters
        self.p_gain = p_gain

        # Set initial conditions
        self.setpoint = 0.

        # Set up publisher
        self.velocity_pub = rospy.Publisher('cmd_vel', Twist, queue_size=10)
        
        # Set up subscribers
        self.setpoint_sub = rospy.Subscriber('setpoint', Float64, self.setpointCallback)
        self.position_sub = rospy.Subscriber('odom', Odometry, self.positionCallback)
        
        # Set up safety timout - will stop robot when timer elapses
        self.timer = rospy.Timer(rospy.Duration(0.1), self.timoutCallback)


    # If this timer runs out, set the robot's velocity to zero
    def timoutCallback(self, msg):
        # Create message
        command = Twist()
        
        # Publish empty message to stop robot
        self.velocity_pub.publish(command)


    # Callback function for incoming setpoint
    def setpointCallback(self, msg):
        self.setpoint = float(msg.data)


    # Callback function for incoming robot position
    def positionCallback(self, msg):
        # Reset the timer
        self.timer.shutdown()
        self.timer.run()

        # Calculate control input
        position = float(msg.pose.pose.position.x)
        error = self.setpoint - position
        input_ = error * self.p_gain

        # Create message
        command = Twist()
        command.linear.x = input_
        
        # Publish message
        self.velocity_pub.publish(command)



#################
# Main function #
#################
def main():
    # Initialize the node
    rospy.init_node('linear_controller', anonymous=True)

    # Create the controller
    controller = LinearController(P_GAIN)
    
    # Wait for messages on topic, go to callback function when new messages arrive.
    rospy.spin()



##########################
# Entry point to program #
##########################
if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: src/linear_controller/src/linear_controller.cpp
#include "linear_controller.h"

// Constructor
LinearController::LinearController(ros::NodeHandle* nh, double p_gain)
{
    // Assign class with reference of nodehandle
    m_nh = *nh;

    // Assign controller parameters
    m_p_gain = p_gain;

    // Set initial conditions
    m_setpoint = 0;

    // Set up publisher
    m_velocity_pub = m_nh.advertise<geometry_msgs::Twist>("cmd_vel", 10);

    // Set up subscribers
    m_setpoint_sub = m_nh.subscribe("setpoint", 10, &LinearController::setpointCallback, this);
    m_position_sub = m_nh.subscribe("odom", 1, &LinearController::positionCallback, this);

    // Set up safety timeout - will stop robot when timer elapses
    m_timer = m_nh.createTimer(ros::Duration(0.1), &LinearController::timoutCallback, this);
}


// If this timer runs out, set the robot's velocity to zero
void LinearController::timoutCallback(const ros::TimerEvent& e)
{
    // Create message
    geometry_msgs::Twist msg;
    
    // Publish empty message to stop robot
    m_velocity_pub.publish(msg);
}


// Callback function for incoming setpoint
void LinearController::setpointCallback(const std_msgs::Float64& data)
{
    // Save setpoint
    m_setpoint = data.data;
}


// Callback function for incoming robot position
void LinearController::positionCallback(const nav_msgs::Odometry& data)
{
    // Reset the timer
    m_timer.stop();
    m_timer.start();

    // Calculate control input
    double position = data.pose.pose.position.x;
    double error = m_setpoint - position;
    double input = error * m_p_gain;

    // Create message
    geometry_msgs::Twist command;
    command.linear.x = input;
    
    // Publish message
    m_velocity_pub.publish(command);
}